File: src/main/java/com/nerdearla/workshop/service/BuyerService.java
package com.nerdearla.workshop.service;

import com.nerdearla.workshop.client.BuyerClient;
import com.nerdearla.workshop.model.Buyer;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

@Service
public class BuyerService {

    private final BuyerClient buyerClient;

    public BuyerService(@Autowired BuyerClient buyerClient) {
        this.buyerClient = buyerClient;
    }

    public Buyer findBuyer(String buyerId) {
        return buyerClient.getById(buyerId);
    }
}

File: src/main/java/com/nerdearla/workshop/service/BuyerService.kt
package com.nerdearla.workshop.service

import com.nerdearla.workshop.dto.buyer.Buyer
import com.nerdearla.workshop.validator.BuyerValidator
import org.springframework.stereotype.Service
import org.springframework.web.reactive.function.client.WebClient

@Service
class BuyerService(
    private val validator: BuyerValidator,
    private val webClient: WebClient
) {
    fun findBuyer(buyerId: String, identification: String, gender: String): Buyer =
        webClient
            .getBuyerById(buyerId)
            .validate(identification, gender)

    private fun Buyer.validate(identification: String, gender: String) =
        also { validator.validate(this, identification, gender) }

    // TODO: Ver si vale la pena armar una capita generica que abstraiga el webclient de los services
    private fun WebClient.getBuyerById(buyerId: String) =
        get()
            .uri { uriBuilder ->
                uriBuilder
                    .path("/buyers/{buyerId}")
                    .build(buyerId)
            }
            .retrieve().bodyToMono(Buyer::class.java).block()!!
}
